packages/p2p/clients/js/index.js

```diff
@@ -0,0 +1,208 @@
+let room
+let events = {}
+let survivingPeer
+let offeringPeer
+let answeringPeer
+let dataChannel
+let connectionStatusCB
+let messageCB
+
+/*** MAIN FUNCTIONS ***/
+
+function connect(userId, uri) {
+  // Set user ID
+  ourId = userId
+
+  server = null
+  theirId = null
+  room = 'empty'
+  survivingPeer = null
+  offeringPeer = null
+  answeringPeer = null
+  dataChannel = null
+  connectionStatusCB = null
+
+  // Connect to server
+  server = new WebSocket(uri)
+
+  // Set up offering peer
+  offeringPeer = new RTCPeerConnection()
+
+  // Set up offering peer's data channel
+  dataChannel = offeringPeer.createDataChannel("dataChannel")
+
+  // Set up answering peer
+  answeringPeer = new RTCPeerConnection()
+
+  // Allow answering peer to accept a data channel from an offer
+  answeringPeer.ondatachannel = handleReceiveDataChannel
+}
+
+function disconnect() {
+  dataChannel.close()
+  leaveRoom()
+  server = null
+  theirId = null
+  events = {}
+  messageCB = null
+}
+
+function listenConnectionStatus(callback) {
+  connectionStatusCB = callback
+  dataChannel.onopen = callback
+  dataChannel.onclose = callback
+}
+
+function listenMessage(callback) {
+  messageCB = callback
+  dataChannel.onmessage = callback
+}
+
+function whenServerReady(callback) {
+  server.on('open', callback)
+}
+
+function whenServerDisconnect(callback) {
+  server.on('close', callback)
+}
+
+function eventName(type, key) {
+  return `${type}-${key}`
+}
+
+function joinRoom(roomName) {
+  room = roomName
+  server.call('joinRoom', [room, ourId])
+}
+
+function leaveRoom() {
+  server.call('leaveRoom', [room, ourId])
+  room = 'empty'
+}
+
+async function rollCall() {
+  const peers = await server.call('rollCall', [room])
+  return peers.filter(id => id != ourId)
+}
+
+function listenOffers(callback) {
+  listenEvent('offer', callback)
+  listenEvent('answer', (e) => {
+    acceptAnswer(e.from, e.data)
+  })
+  listenEvent('candidate', (e) => {
+    addCandidate(e.from, e.data)
+  })
+}
+
+function listenEvent(eventType, callback) {
+  events[eventType] = eventName(eventType, ourId)
+  server.call('addEvent', [events[eventType]])
+  server.subscribe(events[eventType])
+  server.on(events[eventType], async (e) => {
+    callback(e)
+  })
+}
+
+function stopListen(eventType) {
+  server.unsubscribe(events[eventType])
+}
+
+async function sendOffer(id) {
+  // leave the matchmaking system & start the connecting process
+  answeringPeer = null
+  leaveRoom()
+  stopListen('offer')
+  survivingPeer = offeringPeer
+  theirId = id
+
+  // send the offer
+  const offer = await survivingPeer.createOffer()
+  await survivingPeer.setLocalDescription(offer)
+  server.call('triggerEvent', [eventName('offer', theirId), offer, ourId])
+}
+
+async function sendAnswer(id, offer) {
+  // leave the matchmaking system & start the connecting process
+  offeringPeer = null
+  leaveRoom()
+  stopListen('offer')
+  stopListen('answer')
+  survivingPeer = answeringPeer
+  theirId = id
```

Comment:
Why not name it `theirId` in the function deffinition?

Reply:
oh never mind I see. This confusion does highlight that we might not want to use the global scope for all the state variables. Might be better to put them in a single object.

Reply:
I thought about that, and ended up deciding it wasn't worth the extra readability bloat of a state object, but it does seem like it's on the edge. Down to change it if you feel strongly.